### ISSUE TITLE
[css-inline-3] initial-letters changing used, not computed font-siize

### ISSUE THREAD

**Issue by faceless2** [w3c/csswg-drafts#4988]
An element used for `initial-letters` will have its _used_ font-size adjusted based on the algorithm specified in [css-inline-3](https://drafts.csswg.org/css-inline-3/#sizing-initial-letters). However its _computed_ font size remains unchanged.

This is a problem for any child of the "initial-letters" element that refers to the parent font size as part of a style property - for example, if a child sets `font-size: 0.5em`, it will be 50% of the font-size the parent _would have been_, had it not been an initial-letter.

```html
<p style="font-size: 20px; line-height: 1">
 <span style="initial-letters:3">
  A
  <span style="font-size: 0.5em; baseline-shift: 0.5em">
   B
  </span>
 </span>
</p>
```
The letter "A" has a computed font-size of 60px. The letter "B" has a font-size and baseline-shift of 10px, instead of the 30px an author would expect. Anyone creating superscript text as part of an initial letter would encounter this issue.

This issue is migrated from item 6 in https://github.com/w3c/csswg-drafts/issues/4171. There was considerable discussion of this in the January 2020 F2F, which is listed in the [minutes](https://github.com/w3c/csswg-drafts/issues/4171#issuecomment-578112818) however no resolution was reached.

**Comment by faceless2** [comment]
Doing some testing with this, and it's not that simple.

While changing the _computed font size_ would fix the font-size for any superscript text (for example), it would break [example 6](https://drafts.csswg.org/css-inline-3/#example-43e901df) in particular from the css-inline-spec:

```css
p::first-letter {
  initial-letters: 3;
  color: red;
  width: 5em;
  text-align: right;
  margin-left: -5em;
}

p {
  margin-left: 5em;
}
```
Here, the user wants "-5em" to be computed against the specified font size.

**Comment by fantasai** [comment]
Based on the discussion, seems like we should continue to say that the computed font-size is unaffected, and take @emilio’s suggestion of maintaining a multiplier, which will affect the used font size and other effects that key off of the used font size (which is a class of things we'll henceforth need to be more clear about).

I think the only things that need to key off the used font size are going to be anything (other than actual `<length>` values) that keys off the font-size or a character size in the css-text, css-text-decor, and css-inline modules.

Scanning through, looks like that list should include:
- `baseline-shift` percentage values if `line-height` is also a percentage (thus referencing the `font-size` but not via `em` units)
- `letter-spacing` and `word-spacing` percentage values
- `tab-size` number values
- `text-decoration-*` percentage values
- `text-underline-offset` percentage values

Some side effects:
- Need to decouple `line-height` percentages from em units, as they'll behave differently here. (And should have been different from the start but that ship sailed a couple decades ago.)
- If we need another concept of "used font size", e.g. if we introduce per-font multipliers against `font-size` to optically equalize glyphs across a fallback list, then we need to be careful not to conflate terminology with what we're doing here. That kind of effect should not be accounted for here.

**Comment by fantasai** [comment]
Slight complication: https://github.com/w3c/csswg-drafts/issues/2040
Seems that sometimes people want to style the punctuation at the same font size as the body text. Not sure how to make both cases possible.

**Comment by faceless2** [comment]
Invent a "uem" unit for used font-size? Give it a special meaning in initial-letters only?

**Comment by fantasai** [comment]
@faceless2 If we could distinguish inherited font-size vs explicitly set font-size, maybe that would work? Like maybe:
- `initial-letter` sets some inheritable font-size multiplier which is part of the computed value of `font-size` (which is now a tuple, a length and a multiplier, with the length as the basis of `1em`)
- When `font-size` inherits, or is set to a percentage, the multiplier inherits also.
- When `font-size` is set to a `<length>`, the multiplier is reset to 1.

(To help with terminology, we can say the "computed font size" is 1em, but the "used font size" is 1em&times;multiplier.)

**Comment by faceless2** [comment]
I had to read that a few times but I think I see what you're aiming at. So you'd essentially redefine `font-size`, `baseline-shift`, `text-underline-offset` etc. to explicitly include the multiplier aspect of the font-size as part of their calculations, but most other properties (eg `margin-left`) would remain unchanged.

So based on some of the examples linked from #2040, if you wanted `"1st` as your _initial-letter_, with `st` super-scripted and keyed off the used font size, but the leading quote the same size as the paragraph size...

```css
p {
   font-size: 16px;
}
.initial {
    initial-letter: 3 3;
    width: 5em;
    margin-left: -5em;
}
.initial .quote {
    font-size: 16px;
    vertical-align: text-top;
}
sup {
    baseline-shift: 0.4em;
    font-size: 0.5em;
}
```
```html
<p>
 <span class="initial">
  <span class="quote">&lquo;</span>
  1<sup>st</sup>
 </span>
</p>
```
The units used in `sup` are multiplied by our new multiplier, because they're font-relative. But the units in `.quote` are not, because they're absolute.

Have I understood this correctly?

**Comment by fantasai** [comment]
Not quite. `font-size: 0.5em` would reset the multiplier to 1, because it's a `<length>` value. But `font-size: 50%` would maintain the multiplier. This avoids any need to be careful about font-relative lengths generally. :) Only situations where font-relativity is expressed in percentages (line-height, font-size, and various text-* properties) would end up relative to the used font size.
